handler/transport/backend.go

```diff
@@ -120,6 +124,20 @@ func (b *Backend) RoundTrip(req *http.Request) (*http.Response, error) {
 	ctx := context.WithValue(req.Context(), request.BackendURL, req.URL.String())
 	*req = *req.WithContext(ctx)
 
+	if req.Body != nil {
```

Comment:
What about `req.GetBody()`? This method should be populated with a buffer containing the current body bytes.
Since this would apply to all "client" request bodies the token request post data should set the `GetBody()` method too. Prevents the requirement to reset the Body. Maybe a copy to ioutil.Discard and use the resulting `n` would save some memory.
On the other hand, why not setting `req.ContentLength`(or context) together with the body earlier.